src/wolfgame.py

```diff
@@ -1250,96 +1251,503 @@ def stats(cli, nick, chan, rest):
+        # role: [min, max] -- "we may not necessarily know *exactly* how
+        # many of a particular role there are, but we know that there is
+        # between min and max of them"
+        rolecounts = defaultdict(lambda: [0, 0])
+        start_roles = set()
+        orig_roles = {}
+        equiv_sets = {}
+        total_immunizations = 0
+        extra_lycans = 0
+        # Step 1. Get our starting set of roles. This also calculates the maximum numbers for equivalency sets
+        # (sets of roles that are decremented together because we can't know for sure which actually died).
+        for r, v in var.ORIGINAL_ROLES.items():
+            if r in var.TEMPLATE_RESTRICTIONS.keys():
+                continue
+            if len(v) == 0:
+                continue
+            start_roles.add(r)
+            rolecounts[r] = [len(v), len(v)]
+            for p in v:
+                orig_roles[p] = r
+
+        total_immunizations = rolecounts["doctor"][0] * math.ceil(len(var.ALL_PLAYERS) * var.DOCTOR_IMMUNIZATION_MULTIPLIER)
+        if "amnesiac" in start_roles and "doctor" not in var.AMNESIAC_BLACKLIST:
+            total_immunizations += rolecounts["amnesiac"][0] * math.ceil(len(var.ALL_PLAYERS) * var.DOCTOR_IMMUNIZATION_MULTIPLIER)
+
+        extra_lycans = rolecounts["lycan"][0] - min(total_immunizations, rolecounts["lycan"][0])
+
+        equiv_sets["traitor_default"] = rolecounts["traitor"][0] + rolecounts[var.DEFAULT_ROLE][0]
+        equiv_sets["lycan_villager"] = min(rolecounts["lycan"][0], total_immunizations) + rolecounts["villager"][0]
+        equiv_sets["traitor_lycan_villager"] = equiv_sets["traitor_default"] + equiv_sets["lycan_villager"] - rolecounts[var.DEFAULT_ROLE][0]
+        equiv_sets["amnesiac_clone"] = rolecounts["amnesiac"][0] + rolecounts["clone"][0]
+        equiv_sets["amnesiac_clone_cub"] = rolecounts["amnesiac"][0] + rolecounts["clone"][0] + rolecounts["wolf cub"][0]
+        equiv_sets["wolf_fallen"] = 0
+        equiv_sets["fallen_guardian"] = 0
+        if var.TRAITOR_TURNED:
+            equiv_sets["traitor_default"] -= rolecounts["traitor"][0]
+            equiv_sets["traitor_lycan_villager"] -= rolecounts["traitor"][0]
+            rolecounts["wolf"][0] += rolecounts["traitor"][0]
+            rolecounts["wolf"][1] += rolecounts["traitor"][1]
+            rolecounts["traitor"] = [0, 0]
+        # Step 2. Handle role swaps via exchange totem by modifying orig_roles -- the original
+        # roles themselves didn't change, just who has them. By doing the swap early on we greatly
+        # simplify the death logic below in step 3 -- to an outsider that doesn't know any info
+        # the role swap might as well never happened and those people simply started with those roles;
+        # they can't really tell the difference.
+        for a, b in var.EXCHANGED_ROLES:
+            temp = orig_roles[a]
+            orig_roles[a] = orig_roles[b]
+            orig_roles[b] = temp
```

Comment:
This could be written as `(orig_roles[a], orig_roles[b]) = (orig_roles[b], orig_roles[a])` too.